pkg/ccl/crosscluster/logical/lww_row_processor_test.go

```diff
@@ -301,3 +306,165 @@ func BenchmarkLWWInsertBatch(b *testing.B) {
+
+				runner.Exec(t, fmt.Sprintf("INSERT INTO %s VALUES ($1, $2)", tableNameDst), row1...)
+
+				keyValue2 := encoder(timeOneDayForward, row2...)
+				require.NoError(t, insertRow(rp, keyValue2, roachpb.Value{}))
+
+				expectedRows := [][]string{
+					{"1", "row2"},
+				}
+				runner.CheckQueryResults(t, fmt.Sprintf("SELECT * from %s", tableNameDst), expectedRows)
+			})
+
+			t.Run("sql-conflict-incoming-newer", func(t *testing.T) {
+				runner.Exec(t, fmt.Sprintf("SET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled=%t", optimisticInsert))
+				defer runner.Exec(t, "RESET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled")
+				tableNameDst, rp, encoder := setup(t, useKVProc)
+
+				keyValue1 := encoder(timeNow, row1...)
+				require.NoError(t, insertRow(rp, keyValue1, roachpb.Value{}))
+
+				keyValue2 := encoder(timeOneDayForward, row2...)
+				require.NoError(t, insertRow(rp, keyValue2, roachpb.Value{}))
+
+				expectedRows := [][]string{
+					{"1", "row2"},
+				}
+				runner.CheckQueryResults(t, fmt.Sprintf("SELECT * from %s", tableNameDst), expectedRows)
+			})
+
+			t.Run("sql-conflict-incoming-older", func(t *testing.T) {
+				runner.Exec(t, fmt.Sprintf("SET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled=%t", optimisticInsert))
+				defer runner.Exec(t, "RESET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled")
+				tableNameDst, rp, encoder := setup(t, useKVProc)
+
+				keyValue1 := encoder(timeOneDayForward, row1...)
+				require.NoError(t, insertRow(rp, keyValue1, roachpb.Value{}))
+
+				keyValue2 := encoder(timeNow, row2...)
+				require.NoError(t, insertRow(rp, keyValue2, roachpb.Value{}))
+
+				expectedRows := [][]string{
+					{"1", "row1"},
+				}
+				runner.CheckQueryResults(t, fmt.Sprintf("SELECT * from %s", tableNameDst), expectedRows)
+			})
+
+			// From the perspective of the row processor, once the first row is processed, the next incoming event from the
+			// src rangefeed should have a "previous row" that matches the row currently in the cluster. If writes on the src
+			// and dest occur too close together, both tables will attempt to propagate to the other, and the winner of the
+			// conflict will depend on the MVCC timestamp just like the cross cluster write scenario
+			t.Run("outdated-write-conflict", func(t *testing.T) {
+				runner.Exec(t, fmt.Sprintf("SET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled=%t", optimisticInsert))
+				defer runner.Exec(t, "RESET CLUSTER SETTING logical_replication.consumer.try_optimistic_insert.enabled")
+				tableNameDst, rp, encoder := setup(t, useKVProc)
+
+				keyValue1 := encoder(timeNow, row1...)
+				require.NoError(t, insertRow(rp, keyValue1, roachpb.Value{}))
+
+				runner.Exec(t, fmt.Sprintf("UPSERT INTO %s VALUES ($1, $2)", tableNameDst), row2...)
+
+				keyValue3 := encoder(timeOneDayForward, row3...)
```

Comment:
before inserting row3 with `timeOneDayForward`,  could you also try inserting row3 with `time.Now+1` and assert that row 2 wins? In this case, the point is that row3 should have a ts above timeNow and prevVal of row1, but should lose LWW to the local upsert.

Reply:
Isn't this covered by one of the other cases? If row1 and row3 are applied prior to row2, this becomes the cross cluster case I think

Reply:
Actually I see what you're saying

Reply:
Done